scipy/special/_basic.py

```diff
@@ -2791,6 +2791,12 @@ def perm(N, k, exact=False):
 
     """
     if exact:
+        if not (isscalar(N) and isscalar(k)):
+            raise ValueError("`N` and `k` must scalar integers be with `exact=True`.")
```

Comment:
Note that this will be true for 0-dim arrays (though I guess we can live with that). Not necessarily for this PR, but we should also support the array case for `exact=True`, if we do for `exact=False`.

Reply:
Hopefully with the new infrastructure it would be quite easy to write a private ufunc to handle the ``exact=True`` path

Reply:
> Hopefully with the new infrastructure it would be quite easy to write a private ufunc to handle the `exact=True` path

I think the best way to handle the array case for the `exact=True` path in combinatorial functions like this involves scheduling the computations to avoid redundant work, like @rlucas7 did here for `stirling2`:

https://github.com/scipy/scipy/blob/c17b05e397a5e2cf5276920105b082c5ab24bc78/scipy/special/_basic.py#L3305-L3334

Since we need to work with Python ints to avoid overflow, we don't really gain anything by going to native code in a ufunc.

Reply:
Slightly "Yoda" wording on this `ValueError`, but point gets across I guess.